### PR TITLE
Introduce helmer.sh helm manager for tests

### PR DESCRIPTION
**Pull request by kravciak**
Implementation of https://github.com/kubewarden/kubewarden-end-to-end-tests/issues/116

Creating as draft PR, it still needs:
 - clean up commented code
 - test & hook into github workflow

## Most notable features
- move k3d & helm out of makefile
single test run uses github action, makefile, scripts (all of them in sequence) to create cluster & install kubewarden (mixed versions...).
- introduce helmer.sh to keep manage kubewarden during test runs

## helmer.sh

Helm management script:
 - keeps initial installation parameters so we can run follow-up tests with the same setup
 - provides consistent way to modify helm charts without upgrading chart version
 - introduce helpers to install kubewarden stack based on app version
 - allows to test latest images - they are not modified from test files
 - used by makefile targets to install & upgrade kubewarden
```bash
VERSION=1.16.0 CONTROLLER_ARGS="--set image.tag=latest" ./helmer install
CHARTS_LOCATION=./charts DEFAULTS_ARGS="--set recommendedPolicies.enabled=true" ./helmer install

# keeps original values: VERSION, CHARTS_LOCATION, CONTROLLER_ARGS
./helmer set controller --set telemetry.metrics.enabled=False

# reinstall with initial values: VERSION, CHARTS_LOCATION, CONTROLLER_ARGS
helm uninstall kubewarden-defaults -n kubewarden
./helmer reinstall defaults

# Reset helm chart values to initial state
./helmer reset controller
```

## cluster_k3d.sh

Cluster management moved from makefile to cluster_k3d script. Requires 1 action parameter.
```
K3S=1.26 ./cluster_k3d.sh create
./cluster_k3d.sh delete
```

## Other improvements
- common.bash (bats helpers) file split into:
  - kubelib.sh (kubernetes helpers)
  - helpers.sh (kubewarden helpers, require bats)
- policy resource file mover to RESOURCE_DIR/policies
- cleaned up helper functions
wait_for_admiss, apply_admission_... apply_cluster_admi unified into apply_policy and delete_policy

### CODE REVIEW

**Comment by kravciak** [comment]
@jvanz I am removing [generate resources dir](https://github.com/kubewarden/kubewarden-end-to-end-tests/pull/119/files#diff-a1e6d8e62d4bbb2fbee53a531b3fce30c2f646640c6d2aa37e664f3e91761228) script, do you think we might need it?

File: helpers/kubelib.sh [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    local delay=${3:-20}
+    local i
+
+    # export functions into bash -c env (subshell)
+    # export -f helm kubectl
+    # [[ "$cmd" =~ get_metrics ]] && export -f get_metrics
+
+    # Github runner is shared - we must use context for cluster commands
+    #[[ "$cmd" == kubectl* ]] && cmd="${cmd/ / --context $CLUSTER_CONTEXT }"
+    #[[ "$cmd" == helm* ]] && cmd="${cmd/ / --kube-context $CLUSTER_CONTEXT }"
```
[the remark this answers is not in the document]
Comment (kravciak) [reply helpers/kubelib.sh:31]: no, thanks, I will fix that

**Comment by jvanz** [comment]
> ```
> $make clean cluster install basic-end-to-end-tests.bats
> 
> [...]
> 
> basic-end-to-end-tests.bats
>  ✗ [Basic end-to-end tests] Apply pod-privileged policy that blocks CREATE & UPDATE [500]
>    (from function `kubectl' in file tests/../helpers/helpers.sh, line 12,
>     from function `wait_for' in file tests/../helpers/kubelib.sh, line 61,
>     from function `apply_policy' in file tests/../helpers/helpers.sh, line 65,
>     in test file tests/basic-end-to-end-tests.bats, line 16)
>      `apply_policy privileged-pod-policy.yaml' failed
>    clusteradmissionpolicy.policies.kubewarden.io/privileged-pods created
>    error: the server doesn't have a resource type "\"ClusterAdmissionPolicy\""
>  ✓ [Basic end-to-end tests] Patch policy to block only UPDATE operation [120508]
>  ✓ [Basic end-to-end tests] Delete ClusterAdmissionPolicy [290]
>  ✗ [Basic end-to-end tests] Apply mutating psp-user-group AdmissionPolicy [90460]
>    (from function `kubectl' in file tests/../helpers/helpers.sh, line 12,
>     from function `wait_for' in file tests/../helpers/kubelib.sh, line 61,
>     from function `apply_policy' in file tests/../helpers/helpers.sh, line 65,
>     in test file tests/basic-end-to-end-tests.bats, line 45)
>      `apply_policy psp-user-group-policy.yaml' failed
>    policy-server-default-69bd859f98-wfv6j   0/1   ContainerCreating   0     0s     <none>      k3d-k3s-default-server-0   <none>   <none>
>    policy-server-default-69bd859f98-wfv6j   0/1   Running             0     30s     10.42.1.10   k3d-k3s-default-server-0   <none>   <none>
>    audit-scanner-28799400-dpgxl             0/1   ContainerCreating   0     2s      <none>       k3d-k3s-default-server-0   <none>   <none>
>    policy-server-default-89485fbd6-wvxxb    1/1   Terminating   0     3m1s   10.42.0.7    k3d-k3s-default-agent-0    <none>   <none>
>    admissionpolicy.policies.kubewarden.io/psp-user-group created
>    error: the server doesn't have a resource type "\"AdmissionPolicy\""
>  ✓ [Basic end-to-end tests] Launch & scale second policy server [131111]
> 
> 5 tests, 2 failures in 344 seconds
> ```
> 
> We can have a sync talk about this in other communication channels if necessary.

This is a problem in my local `yq` installation. I'm using the wrong one. Thanks @kravciak for the heads up!

File: helpers/helpers.sh [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+function delete_policy {
+    local pfile=$(policypath "$1")
+    kubectl delete --wait -f "$pfile" "${@:2}"
+}
+
+# wait_policies [condition] - at least one policy must exist
+function wait_policies {
+    for chart in ${1:-PolicyActive PolicyUniquelyReachable}; do
+        wait_for --for=condition="$1" admissionpolicies,clusteradmissionpolicies --all -A
```
[the remark this answers is not in the document]
Comment (kravciak) [reply helpers/helpers.sh]: I added policy groups, but we don't have e2e test for them yet, that's next step.

**Comment by kravciak** [comment]
> This is a problem in my local `yq` installation. I'm using the wrong one. Thanks @kravciak for the heads up!

There are 2 implementations of yq, python (kislyuk) and golang (mikefarah).
I wasn't sure which one to choose,
 - yq (mikefarah) is available  in `zypper in yq` and preinstalled on github runners
 - yq (kislyuk) is wrapper for jq, you can use jq syntax and at first glance jq offers more functions

Syntax is different for non-trivial queries - I used mikefarah, but I have no strong opinion here.
I added check for "right" yq version as dependency of `make install` target to avoid confusion.

https://github.com/mikefarah/yq - golang implementation
https://github.com/kislyuk/yq - python wrapper for jq

File: Makefile [hunk 3]
```diff
@@ -1,106 +1,62 @@
...
-
-define generate-versioned-resources-dir =
-	./scripts/generate_resources_dir.sh $(ROOT_RESOURCES_DIR) $(CRD_VERSION)
-endef
+# helmer.sh:
+#   VERSION=[next|prev|v1.17.0-rc2|local] (app version)
+#   REPO_NAME=[kubewarden]
+#   CHARTS_LOCATION=[./dirname|reponame]
+#   LATEST=[1]
+#   CRDS_ARGS, DEFAULTS_ARGS, CONTROLLER_ARGS
```
Comment (jvanz) [inline Makefile:22]: Can we remove this comments?

**Review by jvanz — APPROVED** [review]

File: Makefile (hunk 3 above)
Comment (viccuad) [reply Makefile:22]: I think these are intended to work as inline docs. I'm fine with them.

File: scripts/helmer.sh [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+# Install & Upgrade kubewarden (change chart version)
+
+# Usage: helm_in kubewarden-crds [--params ..]
+helm_in() { helm install --wait --wait-for-jobs --namespace $NAMESPACE "${@:2}" "$1" "$CHARTS_LOCATION/$1"; }
+
+# Install selected $VERSION
+do_install() {
+    echo "Install $VERSION: ${vMap[*]}"
+    # Cert-manager is required by Kubewarden <= v1.16.0 (crds < 1.9.0)
+    if printf '%s\n' "${vMap[crds]}" "1.8.9" | sort -V -C; then
```
Comment (viccuad) [inline scripts/helmer.sh:111]: I wonder if with v1.17.0 out, we can remove the cert-manager install. Doesn't hurt to have it though.

File: scripts/helmer.sh [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+# Configuration of helm versions & values
+
+print_env() {
+    # Main parameters
+    echo NAMESPACE=\"$NAMESPACE\"
+    echo REPO_NAME=\"$REPO_NAME\"
+    echo CHARTS_LOCATION=\"$CHARTS_LOCATION\"
+    echo VERSION=\"$VERSION\"
+    # Extra parameters
+    echo CRDS_ARGS=\"$CRDS_ARGS\"
```
Comment (viccuad) [inline scripts/helmer.sh:50]: Quibble: I would prefer if the args would be read from a temporal values.yaml file per chart, and I could select a different path or manually edit between tests. Sometimes while developing a feature I need a specific setup of Kubewarden, and I can provide that with a big chunk of settings in a values.yaml. Or I need to modify a big chunk of the values.yaml to see how the controller behaves, and would like to run smoke tests too or so.

This is not a blocker and should be done if anything on a different card and PR.

**Review by viccuad — APPROVED** [review]
LGTM! Quite the testcase and helpers refactor!

The new helmer.sh is concise and can be understood, which is a feat given Bash :heart: .

Looking forward to having it merged!

**Review by viccuad — COMMENTED** [review]
Given the amount of Bash and shell code, I think it's time to add a Shellcheck job to the repo CI. ([example](https://github.com/kubewarden/kwctl/blob/ebe5e24fd7a45fe4a04e24b36e4425a14de6bc5a/.github/workflows/ci.yml#L137-L144)).

We can tackle it after merging this PR, it wasn't on the acceptance criteria afterall.

File: scripts/helmer.sh (hunk 5 above)
Comment (kravciak) [reply scripts/helmer.sh:50]: I think this is doable with combination of helmer calls and ARGS.
For example, you can do:
```bash
# Initially
make cluster install CONTROLLER_ARGS="-f values.yaml"

# Later in tests to change it:
helmer set controller -f other_values.yaml
```

File: Makefile (hunk 3 above)
Comment (kravciak) [reply Makefile:22]: yes, it was meant as inline docs, maybe it would be better as README
I would leave it there and maybe move in next iteration..

**Pull request merged**